File: 0101-symmetric-tree/0101-symmetric-tree.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @return {Boolean}
def dfs(left, right)
  return true if !left && !right
  return false if !left || !right
  return false if left.val != right.val
  return dfs(left.left, right.right) && dfs(left.right, right.left)
end
def is_symmetric(root)
  return dfs(root.left, root.right)
end

File: 0101-symmetric-tree/0101-symmetric-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
  return dfs((*root).Left, (*root).Right)
}

func dfs(left *TreeNode, right *TreeNode) bool {
  if left == nil && right == nil {
    return true
  }
  if left == nil || right == nil {
    return false
  }
  if (*left).Val != (*right).Val {
    return false
  }
  return dfs((*left).Left, (*right).Right) && dfs((*left).Right, (*right).Left)
}